extensions/amp-ad-network-mgid-impl/0.1/amp-ad-network-mgid-impl.js

```diff
@@ -0,0 +1,294 @@
+   * @private
+   */
+  getPageParam_() {
+    const counter = Services.urlReplacementsForDoc(
+      this.element
+    )./*OK*/ expandStringSync('COUNTER', undefined, {
+      'COUNTER': true,
+    });
+    return parseInt(counter, 10) % 20;
+  }
+
+  /**
+   * @return {string} Cachebuster for ad request
+   * @private
+   */
+  getCacheBusterParam_() {
+    return (
+      'cbuster=' +
+      Date.now().toString() +
+      Math.floor(Math.random() * 1000000000 + 1)
+    );
+  }
+
+  /**
+   * @return {string} Network information data for ad request
+   * @private
+   */
+  getNetworkInfoParams_() {
+    const params = [];
+    try {
+      const networkInformation =
+        navigator.connection ||
+        navigator.mozConnection ||
+        navigator.webkitConnection;
+
+      if (typeof networkInformation.type != 'undefined') {
+        params.push('nit=' + networkInformation.type);
+      }
+      if (typeof networkInformation.effectiveType != 'undefined') {
+        params.push('niet=' + networkInformation.effectiveType);
+      }
+      if (typeof networkInformation.saveData != 'undefined') {
+        params.push('nisd=' + (networkInformation.saveData ? 1 : 0));
+      }
+    } catch (e) {}
+
+    return params;
+  }
+
+  /**
+   * @return {string} Device pixel ratio info for ad request
+   * @private
+   */
+  getDevicePixelRatioParam_() {
+    let ratio = 1;
+
+    if (typeof window.devicePixelRatio !== 'undefined') {
+      ratio = window.devicePixelRatio;
+    } else if (
+      typeof window.screen.systemXDPI !== 'undefined' &&
+      typeof window.screen.logicalXDPI !== 'undefined' &&
+      window.screen.systemXDPI > window.screen.logicalXDPI
+    ) {
+      ratio = window.screen.systemXDPI / window.screen.logicalXDPI;
+    }
+
+    const isInt = ratio % 1 === 0;
+
+    if (!isInt) {
+      ratio = ratio.toFixed(3);
+    }
+
+    return 'dpr=' + ratio;
+  }
+
+  /**
+   * @return {string} Referrer info for ad request
+   * @private
+   */
+  getPrParam_() {
+    return this.getReferrer_(10).then((referrer) => {
+      const matchDomain = referrer.match(/:\/\/([^\/:]+)/i);
+      return (
+        'pr=' +
+        encodeURIComponent(matchDomain && matchDomain[1] ? matchDomain[1] : '')
+      );
+    });
+  }
+
+  /**
+   * @return {string} Current page url for ad request
+   * @private
+   */
+  getCxurlParam_() {
+    const url = Services.documentInfoForDoc(this.element).canonicalUrl;
+    return 'cxurl=' + encodeURIComponent(url);
+  }
+
+  /**
+   * @return {string} Pageview info for ad request
+   * @private
+   */
+  getPvidParam_() {
+    return Services.documentInfoForDoc(this.element).pageViewId64.then(
+      (pvid) => {
+        return 'pvid=' + pvid;
+      }
+    );
+  }
+
+  /**
+   * @return {string} Muidn info for ad request
+   * @private
+   */
+  getMuidParam_() {
+    return Services.urlReplacementsForDoc(this.element)
+      ./*OK*/ expandStringAsync('CLIENT_ID(muidn)', undefined, {
```

Comment:
did you try just putting this in the ad url directly? I think it should expand there without this extra logic

Reply:
Yes, I've tried it. It works for amp-pixel url, but doesn't for result of getAdUrl.